bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,414 @@
+# Copyright (c) 2023 Boston Dynamics AI Institute Inc.  All rights reserved.
+import atexit
+import collections
+import concurrent.futures
+import dataclasses
+import logging
+import os
+import queue
+import threading
+import typing
+import weakref
+
+import rclpy.executors
+
+
+def classname(obj: typing.Any) -> str:
+    cls = obj.__class__
+    return f"{cls.__module__}.{cls.__qualname__}"
+
+
+class AutoScalingThreadPool(concurrent.futures.Executor):
```

Comment:
> I think all of this logic makes sense, but I am concerned about future debug-ability and getting stuck unexpectedly in the future

> I may just be overly paranoid, but it took me a little bit on paper to walk through things and I probably need to do another pass.

And you're a wise man for it. While adding more tests I uncovered some race conditions :skull_and_crossbones: . I had to change a few things to fix them. PTAL.

> It would be nice to have debug statements that can be enabled/disabled 

I added debug logs all over the place in f014606917ba29ce24713a1f617a0fa9d72bfb9f.